File: Service/BudgetService.cs
using SpendWise.Repository;
using SpendWise.Models;
using SpendWise.DTO.Budget;

namespace SpendWise.Service
{
    public class BudgetService
    {
        private readonly BudgetRepository _budgetRepo;

        public BudgetService(BudgetRepository budgetRepo)
        {
            _budgetRepo = budgetRepo;
        }

        public async Task<BudgetResponseDto> AddBudget(string userId, CreateBudgetDto dto)
        {
            var existing = await _budgetRepo.GetBudget(userId, dto.Month, dto.Year);

            if (existing != null)
            {
                throw new KeyNotFoundException("Budget already exists for this month");
            }

            var budget = new Budget
            {
                UserId = userId,
                BudgetAmount = dto.Amount,
                Month = dto.Month,
                Year = dto.Year,
            };

            budget = await _budgetRepo.AddBudget(budget);

            return new BudgetResponseDto
            {
                Id = budget.Id,
                Amount = budget.BudgetAmount,
                Month = budget.Month,
                Year = budget.Year,
            };
        }

        public async Task<BudgetResponseDto> GetBudget(string userId, int month, int year)
        {
            var budget = await _budgetRepo.GetBudget(userId, month, year);

            if (budget == null)
            {
                throw new KeyNotFoundException("Budget Not Found");
            }

            return new BudgetResponseDto
            {
                Id = budget.Id,
                Amount = budget.BudgetAmount,
                Month = budget.Month,
                Year = budget.Year,
            };
        }

        public async Task<BudgetResponseDto> UpdateBudget(string userId, int budgetId, UpdateBudgetDto dto)
        {
            var updateBudget = await _budgetRepo.UpdateBudget(userId, budgetId, dto);

            return new BudgetResponseDto
            {
                Id = updateBudget.Id,
                Amount = updateBudget.BudgetAmount,
                Month = updateBudget.Month,
                Year = updateBudget.Year,
            };
        }

        public async Task<BudgetResponseDto> DeleteBudget(string userId, int budgetId)
        {
            var deleteBudget = await _budgetRepo.DeleteBudget(userId, budgetId);

            return new BudgetResponseDto
            {
                Id = deleteBudget.Id,
                Amount = deleteBudget.BudgetAmount,
                Month = deleteBudget.Month,
                Year = deleteBudget.Year,
            };
        }
    }
}

File: spendwise-frontend/src/services/budgetService.js
//budgetService.js
import { api } from "./api";

export const budgetService = {
  async createBudget(data, token) {
    return api.post("/budget", data, token);
  },

  async getBudget(month, year, token) {
    try {
      const response = await api.get(`/budget/${month}/${year}`, token);
      return response.budget; // This will be null if no budget exists
    } catch (error) {
      console.error("Failed to fetch budget:", error);
      return null;
    }
  },

  async getAllBudgets(token) {
    return api.get("/budget", token);
  },

  async updateBudget(budgetId, data, token) {
    return api.put(`/budget/${budgetId}`, data, token);
  },

  async deleteBudget(budgetId, token) {
    return api.delete(`/budget/${budgetId}`, token);
  },
};